### PR TITLE
https://issues.redhat.com/browse/ACM-11375

### PR DESCRIPTION
**Pull request by dockerymick**

### CODE REVIEW

File: governance/policy_operator.adoc [hunk 1]
```diff
@@ -41,6 +41,14 @@ If the `remediationAction` set to `inform`, the controller only reports the stat
 | `spec.operatorGroup`
 | Optional
 | By default, if the `operatorGroup` field is not specified, the controller generates an `AllNamespaces` type OperatorGroup in the same namespace as the subscription, if supported. This resource is generated by the operator policy controller.
+//still reviewing
+| `spec.complianceType`
+| Required
+| Specifies whether the described operator should (use value "musthave") or should not (use value "mustnothave) be on the managed cluster.
```
Comment (dockerymick) [inline governance/policy_operator.adoc]:
```suggestion
| Specifies whether the operator should use `musthave` or not use `mustnothave` value on the managed cluster.
```

File: governance/policy_operator.adoc [hunk 2]
```diff
@@ -41,6 +41,14 @@ If the `remediationAction` set to `inform`, the controller only reports the stat
...
+//still reviewing
+| `spec.removalBehavior`
+| Optional 
+| Determines whether certain resource types should be kept or removed when enforcing an OperatorPolicy with `complianceType: mustnothave`; has no effect when `complianceType` is set to `musthave`. The subfield for `operatorGroups` can be set to "Keep" or "DeleteIfUnused" and defaults to "DeleteIfUnused" which will only remove the OperatorGroup if it is not used by any other operators. The subfields for `subscriptions`, and `clusterServiceVersions` can be set to either "Keep" or "Delete", and default to "Delete". The subfield for `customResourceDefinitions` can be set to "Keep" or "Delete", and defaults to "Keep".
```
Comment (dockerymick) [inline governance/policy_operator.adoc]: There is already a row for subscriptions. I recommend that. we move the following statement to line 62: The subfields for `subscriptions`, and `clusterServiceVersions` can be set to either "Keep" or "Delete", and default to "Delete".

It also seems like we need to add another row for `customResourceDefinitions`

```suggestion
| Determines which resource types need to be kept or removed when you enforce an `OperatorPolicy` resource with `complianceType: mustnothave` defined. There is no effect when `complianceType` is set to `musthave`. The subfield for `operatorGroups` can be set to `Keep` or `DeleteIfUnused`. The default value is `DeleteIfUnused` which remove only the `OperatorGroup` resource if it is not used by any other operators. The subfields for `subscriptions`, and `clusterServiceVersions` can be set to either "Keep" or "Delete", and default to "Delete". The subfield for `customResourceDefinitions` can be set to "Keep" or "Delete", and defaults to "Keep". 
```

File: governance/policy_operator.adoc [hunk 3]
```diff
@@ -41,6 +41,14 @@ If the `remediationAction` set to `inform`, the controller only reports the stat
 | `spec.operatorGroup`
 | Optional
 | By default, if the `operatorGroup` field is not specified, the controller generates an `AllNamespaces` type OperatorGroup in the same namespace as the subscription, if supported. This resource is generated by the operator policy controller.
+//still reviewing
```
Comment (dockerymick) [inline governance/policy_operator.adoc]:
```suggestion
```

File: governance/policy_operator.adoc [hunk 4]
```diff
@@ -41,6 +41,14 @@ If the `remediationAction` set to `inform`, the controller only reports the stat
...
+| `spec.complianceType`
+| Required
+| Specifies whether the described operator should (use value "musthave") or should not (use value "mustnothave) be on the managed cluster.
+//still reviewing
```
Comment (dockerymick) [inline governance/policy_operator.adoc]:
```suggestion
```

**Review by dockerymick — COMMENTED** [review]
@JustinKuli will you help by addressing my comments

File: governance/policy_operator.adoc (hunk 1 above)
Comment (JustinKuli) [reply governance/policy_operator.adoc]: I think I phrased this confusingly when I first drafted it.
```suggestion
| Describes the desired state of the operator on the cluster. If set to `musthave`, the policy will have a violation if the operator is missing from the cluster. If set to `mustnothave`, the policy will have a violation if the operator is found on the cluster.
```
Or
```suggestion
| Specifies the desired state of the operator on the cluster. If set to `musthave`, the policy will only be Compliant if the operator is found. If set to `mustnothave`, the policy will be NonCompliant if the operator is found.
```

File: governance/policy_operator.adoc (hunk 2 above)
Comment (JustinKuli) [reply governance/policy_operator.adoc]: My thought was that each row in this table corresponds to a section of the resource definition (YAML). So the `subscriptions` part here is not the same as the `spec.subscription` section we already have a row for. Maybe this example helps:
```yaml
apiVersion: policy.open-cluster-management.io/v1beta1
kind: OperatorPolicy
metadata:
  name: oppol-mustnothave
spec:
  remediationAction: inform
  severity: medium
  complianceType: mustnothave
  subscription:
    channel: stable-3.10
    name: project-quay
    namespace: operator-policy-testns
  versions:
    - quay-operator.v3.10.0
  upgradeApproval: Automatic
  removalBehavior:
    operatorGroups: DeleteIfUnused
    subscriptions: Delete
    clusterServiceVersions: Delete
    customResourceDefinitions: Delete
```

Maybe it also helps to split the subfields as well:
```suggestion
a| Determines which resource types need to be kept or removed when you enforce an `OperatorPolicy` resource with `complianceType: mustnothave` defined. There is no effect when `complianceType` is set to `musthave`. 

- `operatorGroups` can be set to `Keep` or `DeleteIfUnused`. The default value is `DeleteIfUnusued` which only removes the `OperatorGroup` resource if it is not used by any other operators.
- `subscriptions` can be set to `Keep` or `Delete`. The default value is `Delete`.
- `clusterServiceVersions` can be set to `Keep` or `Delete`. The default value is `Delete`.
- `customResourceDefinitions` can be set to `Keep` or `Delete`. The default value is `Keep`. If you set this to `Delete`,  the `CustomResourceDefintion` resources on the managed cluster are removed and can cause data loss.
```

File: governance/policy_operator.adoc (hunk 1 above)
Comment (dockerymick) [reply governance/policy_operator.adoc]:
```suggestion
| Specifies the desired state of the operator on the cluster. If set to `musthave`, the policy is compliant when the operator is found. If set to `mustnothave`, the policy becomes non-compliant when the operator is not found.
```

File: governance/policy_operator.adoc [hunk 5]
```diff
@@ -41,6 +41,12 @@ If the `remediationAction` set to `inform`, the controller only reports the stat
...
 | By default, if the `operatorGroup` field is not specified, the controller generates an `AllNamespaces` type OperatorGroup in the same namespace as the subscription, if supported. This resource is generated by the operator policy controller.
+| `spec.complianceType`
+| Required
+| Specifies the desired state of the operator on the cluster. If set to `musthave`, the policy is compliant when the operator is found. If set to `mustnothave`, the policy becomes non-compliant when the operator is not found.
+| `spec.removalBehavior`
+| Optional 
+| Determines which resource types need to be kept or removed when you enforce an `OperatorPolicy` resource with `complianceType: mustnothave` defined. There is no effect when `complianceType` is set to `musthave`. The subfield for `operatorGroups` can be set to `Keep` or `DeleteIfUnused`. The default value is `DeleteIfUnused` which remove only the `OperatorGroup` resource if it is not used by any other operators. The subfields for `subscriptions`, and `clusterServiceVersions` can be set to either "Keep" or "Delete", and default to "Delete". The subfield for `customResourceDefinitions` can be set to "Keep" or "Delete", and defaults to "Keep".
```
Comment (dockerymick) [inline governance/policy_operator.adoc]:
```suggestion
a| Determines which resource types need to be kept or removed when you enforce an `OperatorPolicy` resource with `complianceType: mustnothave` defined. There is no effect when `complianceType` is set to `musthave`. 
- `operatorGroups` can be set to `Keep` or `DeleteIfUnused`. The default value is `DeleteIfUnusued` which only removes the `OperatorGroup` resource if it is not used by any other operators.
- `subscriptions` can be set to `Keep` or `Delete`. The default value is `Delete`.
- `clusterServiceVersions` can be set to `Keep` or `Delete`. The default value is `Delete`.
- `customResourceDefinitions` can be set to `Keep` or `Delete`. The default value is `Keep`. If you set this to `Delete`,  the `CustomResourceDefintion` resources on the managed cluster are removed and can cause data loss.
```

**Comment by dockerymick** [comment]
@JustinKuli thanks for your patience. I have made some edits based on your feedback. When you have an opportunity, let me know your thoughts

File: governance/policy_operator.adoc [hunk 6]
```diff
@@ -41,6 +41,16 @@ If the `remediationAction` set to `inform`, the controller only reports the stat
 | `spec.operatorGroup`
 | Optional
 | By default, if the `operatorGroup` field is not specified, the controller generates an `AllNamespaces` type OperatorGroup in the same namespace as the subscription, if supported. This resource is generated by the operator policy controller.
+| `spec.complianceType`
+| Required
+| Specifies the desired state of the operator on the cluster. If set to `musthave`, the policy is compliant when the operator is found. If set to `mustnothave`, the policy becomes non-compliant when the operator is not found.
```
Comment (JustinKuli) [inline governance/policy_operator.adoc]: I think we have an accidental double negation here, so either
```suggestion
| Specifies the desired state of the operator on the cluster. If set to `musthave`, the policy is compliant when the operator is found. If set to `mustnothave`, the policy becomes non-compliant when the operator is found.
```
Or
```suggestion
| Specifies the desired state of the operator on the cluster. If set to `musthave`, the policy is compliant when the operator is found. If set to `mustnothave`, the policy is compliant when the operator is not found.
```

**Review by JustinKuli — CHANGES_REQUESTED** [review]
Just one thing, but the rest looks good to me!

**Review by JustinKuli — APPROVED** [review]

**Review by jc-berger — APPROVED** [review]
/lgtm

**Pull request merged**